### PR TITLE
Improve performance for the compressing/decompressing functions

### PR DESCRIPTION
**Pull request by ZehMatt**
From 
image 3.23 seconds
down to
image 2.68 seconds.

Loading a 100 MiB save file, unfortunately I can't upload such huge files here and because its already compressed I can't make it smaller.

### CODE REVIEW

File: src/openrct2/util/Util.cpp [hunk 1]
```diff
@@ -361,53 +362,59 @@ bool UtilGzipCompress(FILE* source, FILE* dest)
...
-    int GZIP_ENCODING = 16;
-    ret = deflateInit2(&strm, Z_DEFAULT_COMPRESSION, Z_DEFLATED, windowBits | GZIP_ENCODING, 8, Z_DEFAULT_STRATEGY);
-    if (ret != Z_OK)
+
+    std::vector<uint8_t> in(InflateDeflateBufferSize);
+    std::vector<uint8_t> out(InflateDeflateBufferSize);
+
+    const int windowBits = 15;
+    const int gzipEncoding = 16;
+    if (const auto ret = deflateInit2(
```
Comment (frutiemax) [inline src/openrct2/util/Util.cpp:377]: Put the deflateInit2 call out of the if condition, only check for ret != Z_OK for cleaner code

File: src/openrct2/util/Util.cpp [hunk 2]
```diff
@@ -361,53 +362,59 @@ bool UtilGzipCompress(FILE* source, FILE* dest)
...
+        strm.next_in = in.data();
         do
         {
-            strm.avail_out = CHUNK;
-            strm.next_out = out;
-            ret = deflate(&strm, flush);
-            if (ret == Z_STREAM_ERROR)
+            strm.avail_out = InflateDeflateBufferSize;
+            strm.next_out = out.data();
+            if (const auto ret = deflate(&strm, flush); ret == Z_STREAM_ERROR)
```
Comment (frutiemax) [inline src/openrct2/util/Util.cpp:401]: same as above

File: src/openrct2/util/Util.cpp [hunk 3]
```diff
@@ -416,46 +423,51 @@ std::vector<uint8_t> Gzip(const void* data, const size_t dataLen)
...
 
     std::vector<uint8_t> output;
+    std::vector<uint8_t> buffer;
+
     z_stream strm{};
     strm.zalloc = Z_NULL;
     strm.zfree = Z_NULL;
     strm.opaque = Z_NULL;
 
+    if (const auto ret = deflateInit2(&strm, Z_DEFAULT_COMPRESSION, Z_DEFLATED, 15 | 16, 8, Z_DEFAULT_STRATEGY); ret != Z_OK)
```
Comment (frutiemax) [inline src/openrct2/util/Util.cpp:433]: same as above

File: src/openrct2/util/Util.cpp [hunk 4]
```diff
@@ -416,46 +423,51 @@ std::vector<uint8_t> Gzip(const void* data, const size_t dataLen)
...
+
         do
         {
-            output.resize(output.size() + nextBlockSize);
             strm.avail_out = static_cast<uInt>(nextBlockSize);
-            strm.next_out = &output[output.size() - nextBlockSize];
-            const auto ret = deflate(&strm, flush);
-            if (ret == Z_STREAM_ERROR)
+            strm.next_out = buffer.data();
+            if (const auto ret = deflate(&strm, flush); ret == Z_STREAM_ERROR)
```
Comment (frutiemax) [inline src/openrct2/util/Util.cpp:459]: same as above

File: src/openrct2/util/Util.cpp [hunk 5]
```diff
@@ -464,46 +476,53 @@ std::vector<uint8_t> Ungzip(const void* data, const size_t dataLen)
...
 
     std::vector<uint8_t> output;
+    std::vector<uint8_t> buffer;
+
     z_stream strm{};
     strm.zalloc = Z_NULL;
     strm.zfree = Z_NULL;
     strm.opaque = Z_NULL;
 
+    if (const auto ret = inflateInit2(&strm, 15 | 16); ret != Z_OK)
```
Comment (frutiemax) [inline src/openrct2/util/Util.cpp:486]: same as above

File: src/openrct2/util/Util.cpp [hunk 6]
```diff
@@ -464,46 +476,53 @@ std::vector<uint8_t> Ungzip(const void* data, const size_t dataLen)
...
+
         do
         {
-            output.resize(output.size() + nextBlockSize);
             strm.avail_out = static_cast<uInt>(nextBlockSize);
-            strm.next_out = &output[output.size() - nextBlockSize];
-            const auto ret = inflate(&strm, flush);
-            if (ret == Z_STREAM_ERROR)
+            strm.next_out = buffer.data();
+            if (const auto ret = inflate(&strm, flush); ret == Z_STREAM_ERROR)
```
Comment (frutiemax) [inline src/openrct2/util/Util.cpp:513]: same

**Review by frutiemax — COMMENTED** [review]
Statement in if condition is a bit evil, I would declare a local variable for the return code and check against it in the if conditions.

**Comment by ZehMatt** [comment]
> Statement in if condition is a bit evil, I would declare a local variable for the return code and check against it in the if conditions.

What exactly is evil about it? This is clearly defined by the standard, see https://en.cppreference.com/w/cpp/language/if , the advantage of this is you are having less active variables in the scope. See https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0305r0.html for the reasons why it was added

File: src/openrct2/util/Util.cpp [hunk 7]
```diff
@@ -361,53 +362,59 @@ bool UtilGzipCompress(FILE* source, FILE* dest)
...
-    int windowBits = 15;
-    int GZIP_ENCODING = 16;
-    ret = deflateInit2(&strm, Z_DEFAULT_COMPRESSION, Z_DEFLATED, windowBits | GZIP_ENCODING, 8, Z_DEFAULT_STRATEGY);
-    if (ret != Z_OK)
+
+    std::vector<uint8_t> in(InflateDeflateBufferSize);
+    std::vector<uint8_t> out(InflateDeflateBufferSize);
+
+    const int windowBits = 15;
+    const int gzipEncoding = 16;
```
Comment (tupaschoal) [inline src/openrct2/util/Util.cpp:376]:
```suggestion
    constexpr intwindowBits = 15;
    constexpr int gzipEncoding = 16;
```
Might as well make them `constexpr`?

File: src/openrct2/util/Util.cpp [hunk 8]
```diff
@@ -361,53 +362,59 @@ bool UtilGzipCompress(FILE* source, FILE* dest)
...
+    if (const auto ret = deflateInit2(
+            &strm, Z_DEFAULT_COMPRESSION, Z_DEFLATED, windowBits | gzipEncoding, 8, Z_DEFAULT_STRATEGY);
```
Comment (tupaschoal) [inline src/openrct2/util/Util.cpp:378]: Maybe extract `8` to a variable like:
```c++
constexpr int memoryUsageLevel = 8;
```

**Review by tupaschoal — COMMENTED** [review]
I'm ok with the `ifs`. Is the performance boost mostly due to the `reserve`s or did increasing the chunk size also make it better?

File: src/openrct2/util/Util.cpp (hunk 1 above)
Comment (ZehMatt) [reply src/openrct2/util/Util.cpp:377]: There are multiple places in the code base that uses this, we should continue doing so.

**Comment by ZehMatt** [comment]
> I'm ok with the `ifs`. Is the performance boost mostly due to the `reserve`s or did increasing the chunk size also make it better?

A few factors that contribute to the performance increase, chunk size was increased so it can directly encode/decode more into the buffer, using vector resize each time will not have extra capacity  where using insert is the same as push_back and there will be likely spare room for the next insert, the reserve call should also help avoid some allocations since vector growth is typically something like 1.5x so starting empty would have a bunch of allocations.

**Comment by ZehMatt** [comment]
Found a park that can't load, need to investigate that first and probably a good idea to write tests for this

**Pull request closed without merging**